samza-api/src/main/java/org/apache/samza/storage/kv/ClosableIterator.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.storage.kv;
+
+import java.util.Iterator;
+
+/**
+ * An iterator that can be closed.
+ *
+ * <p> Implement close to free resources assigned to the iterator such as open file handles, persistent state etc.
+ *
+ * @param <V> the type of value returned by this iterator
+ */
+public interface ClosableIterator<V> extends Iterator<V> {
```

Comment:
Seems like it would be better to leverage the java Closeable interface so we can support AutoCloseable for try-with-resources
https://docs.oracle.com/javase/7/docs/api/java/lang/AutoCloseable.html

Reply:
1. While the `AutoClosable` idiom is useful for file handles, it expects the caller to catch the checked `Exception` each time they call `iterator.close()`. Almost always, In Samza we 'd want to shutdown when there's an exception during `iterator.close()`, So, a runtime exception makes more sense.

2. Furthermore,  `iterator.close()` on our existing `KVStore` API does not throw any checked `Exception`. I'd prefer to lean on the side of consistency, and refactor it later if there's enough consensus. Happy to discuss offline, and follow-up!